primitives/blockchain/src/header_metadata.rs

```diff
@@ -179,9 +179,13 @@ pub struct TreeRoute<Block: BlockT> {
 impl<Block: BlockT> TreeRoute<Block> {
 	/// Creates a new `TreeRoute`.
 	///
-	/// It is required that `pivot >= route.len()`, otherwise it may panics.
-	pub fn new(route: Vec<HashAndNumber<Block>>, pivot: usize) -> Self {
-		TreeRoute { route, pivot }
+	/// To preserve the structure safety invariats it is required that `pivot < route.len()`.
+	pub fn new(route: Vec<HashAndNumber<Block>>, pivot: usize) -> Result<Self, String> {
+		if pivot < route.len() {
+			Ok(TreeRoute { route, pivot })
+		} else {
+			Err("TreeRoute pivot should be less than route length".into())
```

Comment:
Not a big deal, but if you're returning a `String` as an error and not a `&'static str` it might be nice to take advantage of that and print out a more detailed error message. (:

Reply:
I think here is better to go with static str. The string is statically allocated in the binary anyway so indeed doesn't have any sense returning it as a `String`.

Reply:
Yep. That works too!